src/Core/Resolvers/SqlMutationEngine.cs

```diff
@@ -367,135 +376,383 @@ public class SqlMutationEngine : IMutationEngine
-                    if (!isUpdateResultSet ||
-                        (sqlMetadataProvider.GetDatabaseType() is DatabaseType.PostgreSQL &&
-                        PostgresQueryBuilder.IsInsert(resultRow)))
+                    Dictionary<string, object?> resultRow = upsertOperationResultSetRow.Columns;
+
+                    // For all SQL database types, the IS_UPDATE_RESULT_SET field is added to the resultset to indicate if the resultant operation was an update or insert.
+                    // For MsSQL and MySQL, the result set will contain one entry when the resultant operation is update and two entries incase of insert.
+                    // For insert operation, the first result row will not contain any columns. This is used to determine that the resultant operation was an update.
+                    // In this case, IS_UPDATE_RESULT_SET field is added to the result set.
+                    // However, this logic is inapplicable for PostgreSQL, as the result set of an upsert operation will contain only one entry when both update or an insert takes place.
+                    // The result set will contain a field (___upsert_op___) that indicates whether the resultant operation was an update or an insert. So, the value present in this field
+                    // is used to determine whether the upsert resulted in an update.
+                    if (sqlMetadataProvider.GetDatabaseType() is DatabaseType.PostgreSQL)
                     {
-                        string primaryKeyRoute = ConstructPrimaryKeyRoute(context, resultRow, sqlMetadataProvider);
-                        // location will be updated in rest controller where httpcontext is available
-                        return new CreatedResult(location: primaryKeyRoute, OkMutationResponse(resultRow).Value);
+                        hasPerformedUpdate = !PostgresQueryBuilder.IsInsert(resultRow);
                     }
 
-                    // Valid REST updates return OkObjectResult
-                    return OkMutationResponse(resultRow);
-                }
-            }
-            else
-            {
-                DbResultSetRow? mutationResultRow = null;
+                    // When read permissions is configured without database policy, a subsequent select query will not be executed.
+                    // However, the read action could have include and exclude fields configured. To honor that configuration setup,
+                    // any additional fields that are present in the response are removed.
+                    if (isReadPermissionConfiguredForRole && !isDatabasePolicyDefinedForReadAction)
+                    {
+                        IEnumerable<string> allowedExposedColumns = _authorizationResolver.GetAllowedExposedColumns(context.EntityName, roleName, EntityActionOperation.Read);
+                        foreach (string columnInResponse in resultRow.Keys)
```

Comment:
Since insert operation will have 2 result sets, when do we move the resultRow to the second set?

Reply:
Insert operation will contain two result sets.  

The logic for handling - moving to the second result set and fetching the result row is at `QueryExecutor.GetMultipleResultSetsIfAnyAsync()` or `MsSqlQueryExecutor.GetMultipleResultSetsIfAnyAsync()`. This returns only the second result set [Incase of update, the first result set is returned].

So, when the flow reaches back `SqlMutationEngine`, the result set returned will be a valid one that contains information about the record inserted/updated.

Reply:
Why would insert operation contain two result sets? Just to clarify here:
Two result sets (at max) are present only for upsert operations in which:
1. First result set holds the count of rows with given PK (0/1). This result set is always present.
2. The second result set is for the update/insert operation. In case neither update/insert was performed, there should be no second result set.